File: java/Walker.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.concurrent.ThreadLocalRandom;

public class Walker {
	ArrayList<Point> marbles = new ArrayList<>();
	HashSet<Point> visited = new HashSet<>();

	public Walker(int n, int k) {
		Point location = new Point(0, 0);
		this.visited.add(location);
		if (n == 0 || k == 0) return;
		while (!trapped(location)) {
			Point desired = validNeighbor(location);
			if (this.visited.size() % n == 0) {
				this.marbles.add(desired);
			}
			this.visited.add(desired);
			location = desired;
			if (visited.size() > n * k) break;
		}
	}

	public int intersections() {
		return (new BestLine(this.marbles)).intersections();
	}

	public int steps() {
		return this.visited.size();
	}

	private boolean trapped(Point p) {
		return this.visited.contains(p.up()) &&
				this.visited.contains(p.right()) &&
				this.visited.contains(p.down()) &&
				this.visited.contains(p.left());
	}

	private Point validNeighbor(Point p) {
		Point neighbor = randomNeighbor(p);
		while (this.visited.contains(neighbor)) {
			neighbor = randomNeighbor(p);
		}
		return neighbor;
	}

	private static Point randomNeighbor(Point p) {
		int direction = ThreadLocalRandom.current().nextInt(0, 4);
		switch (direction) {
			case 0: return p.up();
			case 1: return p.right();
			case 2: return p.down();
			default: return p.left();
		}
	}
}

File: kotlin/Walker.kt
class Walker(n: Int, k: Int) {
	var marbles: MutableList<Point>
	var visited: MutableSet<Point>

	init {
		var location = Point(0, 0)
		marbles = mutableListOf<Point>()
		visited = mutableSetOf<Point>(location)
		if (n != 0 && k != 0) {
			while (!trapped(location)) {
				val desired = valid_neighbor(location)
				if (visited.size % n == 0) {
					marbles.add(desired)
				}
				visited.add(desired)
				location = desired
				if (visited.size > n * k) {
					break;
				}
			}
		}
	}

	fun steps(): Int {
		return visited.size
	}

	fun intersections(): Int {
		return BestLine(marbles).intersections()
	}

	private fun trapped(p: Point): Boolean {
		return visited.contains(p.up()) &&
				visited.contains(p.right()) &&
				visited.contains(p.down()) &&
				visited.contains(p.left())
	}

	private fun valid_neighbor(p: Point): Point {
		var neighbor = random_neighbor(p)
		while (visited.contains(neighbor)) {
			neighbor = random_neighbor(p)
		}
		return neighbor
	}

	private fun random_neighbor(p: Point): Point {
		when ((0..3).random()) {
			0 -> return p.up()
			1 -> return p.right()
			2 -> return p.down()
			else -> return p.left()
		}
	}
}
